python/lsst/pipe/base/pipeline.py

```diff
@@ -879,27 +886,51 @@ def makeDatasetTypesSet(connectionType: str, freeze: bool = True) -> NamedValueS
                         )
 
                     if registryDatasetType and datasetType != registryDatasetType:
-                        try:
-                            # Explicitly check for storage class just to make
-                            # more specific message.
-                            _ = datasetType.storageClass
-                        except KeyError:
+                        # The dataset types differ but first check to see if
+                        # they are compatible before raising.
+                        if is_input:
+                            # This DatasetType must be compatible on get.
+                            is_compatible = datasetType.is_compatible_with(registryDatasetType)
+                        else:
+                            # Has to be able to be converted to expect type
+                            # on put.
+                            is_compatible = registryDatasetType.is_compatible_with(datasetType)
+                        if is_compatible:
+                            # For inputs we want the pipeline to use the
+                            # pipeline definition, for outputs it should use
+                            # the registry definition.
+                            if not is_input:
+                                datasetType = registryDatasetType
+                            _LOG.warning(
```

Comment:
Should it be `debug`?

Reply:
Yes. I meant to do that -- I only had them warning to make it easy to spot them in debugging.